test/framework/src/main/java/org/elasticsearch/test/rest/TestFeatureService.java

```diff
@@ -22,18 +23,31 @@ class TestFeatureService {
     private final Predicate<String> historicalFeaturesPredicate;
     private final Set<String> clusterStateFeatures;
 
-    TestFeatureService(List<? extends FeatureSpecification> specs, Collection<Version> nodeVersions, Set<String> clusterStateFeatures) {
-
+    TestFeatureService(
+        boolean hasHistoricalFeaturesInformation,
+        List<? extends FeatureSpecification> specs,
+        Collection<Version> nodeVersions,
+        Set<String> clusterStateFeatures
+    ) {
         var minNodeVersion = nodeVersions.stream().min(Version::compareTo);
         var featureData = FeatureData.createFromSpecifications(specs);
         var historicalFeatures = featureData.getHistoricalFeatures();
         var allHistoricalFeatures = historicalFeatures.lastEntry() == null ? Set.of() : historicalFeatures.lastEntry().getValue();
+
+        var errorMessage = hasHistoricalFeaturesInformation
+            ? "Check the feature has been added to the correct FeatureSpecification in the relevant module or, if this is a "
+                + "legacy feature used only in tests, to a test-only FeatureSpecification"
+            : "This test seems to run on the legacy test plugins; historical features from production code will not be available."
+                + " You need to port the test to the new test plugins in order to use historical features from production code."
+                + " If this is a legacy feature used only in tests, you can add it to a test-only FeatureSpecification";
         this.historicalFeaturesPredicate = minNodeVersion.<Predicate<String>>map(v -> featureId -> {
-            assert allHistoricalFeatures.contains(featureId) : "Unknown historical feature " + featureId;
+            assert allHistoricalFeatures.contains(featureId)
+                : String.format(Locale.ROOT, "Unknown historical feature %s: %s", featureId, errorMessage);
```

Comment:
`Strings.format` does the locale for you

Reply:
Thanks for letting me know, that's better!